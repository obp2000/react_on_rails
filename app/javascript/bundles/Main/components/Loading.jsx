import React from 'react'
import loading_spinner from './loading_spinner.gif'

const Loading = () => {
    return (
        <tr>
            <td>
                <img src={loading_spinner} width="20px"/>
            </td>
        </tr>
    )
}

export default Loading
