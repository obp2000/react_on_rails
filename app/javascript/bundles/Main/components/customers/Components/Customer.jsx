import PropTypes from 'prop-types'
import React from 'react'
import NewCustomer from '../Containers/NewCustomer'

export default class Customer extends React.Component {
    constructor(props) {
        super(props)
    }
    componentDidMount() {
        this
            .props
            .fetchCustomer()
    }
    render() {
        return <NewCustomer/>
    }
    static propTypes = {
        fetchCustomer: PropTypes.func.isRequired
    }
}
