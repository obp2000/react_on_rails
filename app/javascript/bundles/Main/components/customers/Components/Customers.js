import PropTypes from 'prop-types'
import React from 'react'
import Container from '../Containers/Customers'

export default class Customers extends React.Component {
  constructor(props) {
    super(props)
  }
  componentDidMount() {
    this
      .props
      .fetchCustomers()
  }
  componentDidUpdate() {
    this
      .props
      .fetchCustomers()
  }
  render() {
    return <Container/>
  }

  static propTypes = {
    fetchCustomers: PropTypes.func.isRequired
  }
}
