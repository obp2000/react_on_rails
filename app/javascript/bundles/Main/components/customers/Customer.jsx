import PropTypes from 'prop-types'
import React from 'react'
import {Link} from 'react-router-dom'
import DeleteCustomerButton from './Containers/DeleteCustomerButton'

const Customer = ({
    id,
    nick,
    name,
    created_at,
    updated_at,
}) => <tr>
    <td>{id}</td>
    <td>{nick}</td>
    <td>{name}</td>
    <td>{created_at}</td>
    <td>{updated_at}</td>
    <td>
        <Link to={"/customers/" + id} className="btn btn-outline-primary btn-sm">Редактировать</Link>
    </td>
    <td>
        <DeleteCustomerButton id={id}/>
    </td>
</tr>

Customer.propTypes = {
    id: PropTypes.number,
    nick: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    address: PropTypes.string,
    created_at: PropTypes.string.isRequired,
    updated_at: PropTypes.string.isRequired,
}

Customer.defaultProps = {
    id: 0,
    nick: '',
    name: '',
    address: '',
    created_at: '',
    updated_at: ''
}

export default Customer