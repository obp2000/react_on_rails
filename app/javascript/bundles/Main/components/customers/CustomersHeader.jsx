import React from 'react'
import {Link} from 'react-router-dom'

const CustomersHeader = () => {
    return <thead className="thead-light">
        <tr>
            <th scope="col">Id</th>
            <th scope="col">Ник</th>
            <th scope="col">ФИО</th>
            <th scope="col">Создан</th>
            <th scope="col">Обновлен</th>
            <th scope="col">
                <Link to="/customers/new" className="btn btn-outline-primary btn-sm">Новый</Link>
            </th>
            <th scope="col"></th>
        </tr>
    </thead>
}

export default CustomersHeader