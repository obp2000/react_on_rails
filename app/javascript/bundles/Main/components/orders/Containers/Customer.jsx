import {connect} from 'react-redux'
import {formValueSelector} from 'redux-form'
import Customer from '../Customer'

const mapStateToProps = (state) => {
    const customer = formValueSelector('order')(state, 'customer')
    return customer
}

export default connect(mapStateToProps)(Customer)