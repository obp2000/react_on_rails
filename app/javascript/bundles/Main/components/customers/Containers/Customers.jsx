import {connect} from 'react-redux'
import Customers from '../Customers'

const mapStateToProps = ({
  customers: {
    customers,
    totalCount,
    isFetching
  }
}) => {
  return {customers, totalCount, isFetching}
}

export default connect(mapStateToProps)(Customers)

