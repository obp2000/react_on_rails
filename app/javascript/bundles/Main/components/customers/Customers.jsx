import PropTypes from 'prop-types'
import React from 'react'

import Loading from '../Loading'
import CustomersHeader from './CustomersHeader'
import Customer from './Customer'

const Customers = ({customers, totalCount, isFetching}) => <div>
  <h3>Покупатели ({totalCount})</h3>
  <table className="table table-sm table-striped table-bordered table-hover">
    <CustomersHeader/>
    <tbody>
      {!isFetching
        ? customers.map((customer, index) => {
          return <Customer {...customer} key={index}/>;
        })
        : <Loading/>}
    </tbody>
  </table>
</div>

Customers.propTypes = {
  customers: PropTypes.array.isRequired,
  totalCount: PropTypes.number,
  isFetching: PropTypes.bool
}

Customers.defaultProps = {
  customers: []
}

export default Customers
