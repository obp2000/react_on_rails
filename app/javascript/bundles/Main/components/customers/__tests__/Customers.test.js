import React from 'react'
import Enzyme, {shallow} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import CustomersHeader from '../CustomersHeader'
import Customers from '../Customers'
import Customer from '../Customer'
import Loading from '../../Loading'
jest.mock('../../loading_spinner.gif', () => 'loading_spinner.gif')

Enzyme.configure({adapter: new Adapter()})

function setup(isFetching = false) {
  const props = {
    customers: ['customer1', 'customer2'],
    totalCount: 2,
    isFetching
  }

  const customers = shallow(<Customers {...props}/>)

  return {customers, props}
}

describe('components', () => {
  describe('Customers', () => {
    it('should render title and header', () => {
      const {customers, props} = setup(false)
      const title = customers.find('h3')
      expect(title.length).toBe(1)
      expect(title.text()).toBe(`Покупатели (${props.totalCount})`)
      const header = customers.find(CustomersHeader)
      expect(header.length).toBe(1)
    })
    it('should render list of customers and not render loading after fetching ', () => {
      const {customers, props} = setup(false)
      const customersList = customers.find(Customer)
      expect(customersList.length).toBe(2)
      const loading = customers.find(Loading)
      expect(loading.length).toBe(0)
    })
    it('should render loading and not render list of customers in fetching', () => {
      const {customers, props} = setup(true)
      const loading = customers.find(Loading)
      expect(loading.length).toBe(1)
      const customersList = customers.find(Customer)
      expect(customersList.length).toBe(0)
    })
  })
})
