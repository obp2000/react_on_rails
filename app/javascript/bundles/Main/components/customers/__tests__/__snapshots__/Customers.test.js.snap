// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Customers 1`] = `
<div>
  <h3>
    Покупатели (
    )
  </h3>
  <table
    className="table table-sm table-striped table-bordered table-hover"
  >
    <thead
      className="thead-light"
    >
      <tr>
        <th
          scope="col"
        >
          Id
        </th>
        <th
          scope="col"
        >
          Ник
        </th>
        <th
          scope="col"
        >
          ФИО
        </th>
        <th
          scope="col"
        >
          Создан
        </th>
        <th
          scope="col"
        >
          Обновлен
        </th>
        <th
          scope="col"
        >
          <a
            className="btn btn-outline-primary btn-sm"
            href="#/customers/new"
            onClick={[Function]}
          >
            Новый
          </a>
        </th>
        <th
          scope="col"
        />
      </tr>
    </thead>
    <tbody />
  </table>
</div>
`;
